vllm_v1/core/scheduler.py

```diff
@@ -0,0 +1,367 @@
+        scheduled_running_reqs: List[Request] = []
+        preempted_reqs: List[Request] = []
+
+        req_to_new_block_ids: Dict[str, List[int]] = {}
+        num_scheduled_tokens: Dict[str, int] = {}
+        token_budget = self.max_num_scheduled_tokens
+
+        # First, schedule the RUNNING requests.
+        req_index = 0
+        while req_index < len(self.running):
+            if token_budget == 0:
+                break
+
+            request = self.running[req_index]
+            num_tokens = request.num_tokens - request.num_computed_tokens
+            num_tokens = min(num_tokens, token_budget)
+            assert num_tokens > 0
+
+            while True:
+                start = time.time()
+                new_block_ids = self.kv_cache_manager.append_slots(
+                    request, num_tokens)
+                end = time.time()
+                self.cum += (end - start)
+                if new_block_ids is None:
+                    # The request cannot be scheduled.
+                    # Preempt the lowest-priority request.
+                    preempted_req = self.running.pop()
+                    self.kv_cache_manager.free(preempted_req)
+                    preempted_req.status = RequestStatus.PREEMPTED
+                    preempted_req.num_computed_tokens = 0
+
+                    self.waiting.appendleft(preempted_req)
+                    preempted_reqs.append(preempted_req)
+                    if preempted_req == request:
+                        # No more request to preempt.
+                        break
+                else:
+                    # The request can be scheduled.
+                    scheduled_running_reqs.append(request)
+
+                    req_to_new_block_ids[request.request_id] = new_block_ids
+                    num_scheduled_tokens[request.request_id] = num_tokens
+                    token_budget -= num_tokens
+                    req_index += 1
+                    break
+
+        start = time.time()
+        # Next, schedule the WAITING requests.
+        while self.waiting:
+            if preempted_reqs:
+                break
+            if len(self.running) == self.max_num_running_reqs:
+                break
+            if token_budget == 0:
+                break
+
+            request = self.waiting[0]
+            # Get already-cached tokens.
+            computed_block_ids = self.kv_cache_manager.get_computed_blocks(
+                request)
+            # NOTE(woosuk): Since incomplete blocks are not eligible for
+            # sharing, `num_computed_tokens` is always a multiple of
+            # `block_size`.
+            num_computed_tokens = len(computed_block_ids) * self.block_size
+            # Number of tokens to be scheduled.
+            # We use `request.num_tokens` instead of `request.num_prompt_tokens`
+            # to consider the resumed requests, which have output tokens.
+            num_tokens = request.num_tokens - num_computed_tokens
```

Comment:
That's a great question. We actually never do the full cache hit, because in that case we don't have the "logits" we can sample the next token from. We always leave at least 1 block for the input.

* This is the implementation in vllm V0's prefix caching, which I'd like to keep for this re-arch.